adapters/tpmn/tpmn.go

```diff
@@ -0,0 +1,159 @@
+	URI string
+}
+
+// MakeRequests makes the HTTP requests which should be made to fetch bids from TpmnBidder.
+func (rcv *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	validImps, errs := getValidImpressions(request, reqInfo)
+	if len(validImps) == 0 {
+		return nil, errs
+	}
+
+	request.Imp = validImps
+
+	requestBodyJSON, err := json.Marshal(request)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	return []*adapters.RequestData{{
+		Method:  http.MethodPost,
+		Uri:     rcv.URI,
+		Body:    requestBodyJSON,
+		Headers: headers,
+	}}, errs
+}
+
+// getValidImpressions validate imps and check for bid floor currency. Convert to EUR if necessary
+func getValidImpressions(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]openrtb2.Imp, []error) {
+	var errs []error
+	var validImps []openrtb2.Imp
+
+	for _, imp := range request.Imp {
+		if err := preprocessBidFloorCurrency(&imp, reqInfo); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if err := preprocessExtensions(&imp); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		validImps = append(validImps, imp)
+	}
+	return validImps, errs
+}
+
+func preprocessExtensions(imp *openrtb2.Imp) error {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	var tpmnBidderExt openrtb_ext.ExtImpTpmn
+	if err := json.Unmarshal(bidderExt.Bidder, &tpmnBidderExt); err != nil {
+		return &errortypes.BadInput{
+			Message: "Wrong TPMN bidder ext: " + err.Error(),
+		}
+	}
+
+	return nil
+}
+
+func preprocessBidFloorCurrency(imp *openrtb2.Imp, reqInfo *adapters.ExtraRequestInfo) error {
+	// we expect every currency related data to be EUR
+	if imp.BidFloor > 0 && strings.ToUpper(imp.BidFloorCur) != "USD" && imp.BidFloorCur != "" {
+		if convertedValue, err := reqInfo.ConvertCurrency(imp.BidFloor, imp.BidFloorCur, "USD"); err != nil {
+			return err
+		} else {
+			imp.BidFloor = convertedValue
+		}
+	}
+	imp.BidFloorCur = "USD"
+	return nil
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, _ *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	switch responseData.StatusCode {
+	case http.StatusNoContent:
+		return nil, nil
+	case http.StatusBadRequest:
+		return nil, []error{errors.New("bad request from publisher")}
+	case http.StatusOK:
+		break
+	default:
+		return nil, []error{fmt.Errorf("unexpected response status code: %v", responseData.StatusCode)}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{fmt.Errorf("bid response unmarshal: %v", err)}
+	}
+
```

Comment:
https://github.com/prebid/prebid-server/blob/eb338b6c4c68b778cf103bb7028dd1245ab1aee9/adapters/bidder.go#L40-L41

please use `errortypes.BadInput` and `errortypes.BadServerResponse ` as mentioned above

Reply:
update commit : 
[8cec466](https://github.com/prebid/prebid-server/pull/2947/commits/8cec4663d47621ff2654b38692c5391800b7b3de)